### ISSUE TITLE
build your project under msys2

### ISSUE THREAD

**Issue by asmwarrior** [PBfordev/wxOpenCVCameras#2]
Hi, I have fork this project, and build this project under msys2.

You can see my changes here:  https://github.com/asmwarrior/wxOpenCVCameras/commits/master

Bacially, I have to tweak the CMakefiles.txt.

BTW: your project is great!

**Comment by PBfordev** [comment]
I know little about CMake and nothing about MSYS2 but TBH, these changes do not seem correct to me.

Firstly, they are hard-coded, not taking wxWidgets version and CMake Generator used into account, which will make the project unbuildable unless one is using MSYS2 and wxWidgets 3.1...

Secondly, the missing manifest should be hopefully fixed in 3.1.6+ by [this](https://github.com/wxWidgets/wxWidgets/commit/77d892612640a23c17debdac579c4afcbb2b4afd#diff-21645892d42128ec76f4115b761da2c2b4a7a8a9d04ac71152534ec11c70a3f1).

Thirdly, I also do not understand that you say that CMake module FindwxWidgets cannot find `${wxWidgets_ROOT_DIR}`. If so, why not just tell it where it is? 

Lastly, does the path to the resource file really has to be hard-coded like this `"${wxWidgets_LIBRARY_DIRS}/../include/wx-3.1/wx/msw/wx.rc"`. Can you not use a CMake variable that points to the currently used include dir instead?

**Comment by asmwarrior** [comment]
> Lastly, does the path to the resource file really has to be hard-coded like this `"${wxWidgets_LIBRARY_DIRS}/../include/wx-3.1/wx/msw/wx.rc"`. Can you not use a CMake variable that points to the currently used include dir instead?

I can wrote something like: 

`list(APPEND SOURCES "F:/msys2/mingw64/include/wx-3.1/wx/msw/wx.rc")`

But I don't want to write hard-coded file path in the CMakeFiles.txt file. Because it is not relocatable for others.


> Thirdly, I also do not understand that you say that CMake module FindwxWidgets cannot find ${wxWidgets_ROOT_DIR}. If so, why not just tell it where it is?

Yes, I can tell CMake where the ${wxWidgets_ROOT_DIR} is, but people need to pass this variable to CMake when they invoke the CMake command.

**Comment by asmwarrior** [comment]
`set(wxWidgets_CONFIG_EXECUTABLE "wx-config-3.1")`

This code is not needed of we can set an environment variable `WX_CONFIG`, see below:

~~~~
zyh@WINMICR-0PLE9E2 MINGW64 /e/code/wxOpenCVCameras/build
# export WX_CONFIG=${MINGW_PREFIX}/bin/wx-config-3.1

zyh@WINMICR-0PLE9E2 MINGW64 /e/code/wxOpenCVCameras/build
# echo $WX_CONFIG
/mingw64/bin/wx-config-3.1

zyh@WINMICR-0PLE9E2 MINGW64 /e/code/wxOpenCVCameras/build
# cmake ../
-- F:/msys2/mingw64/share/cmake/Modules/FindwxWidgets.cmake(176): wxWidgets_FIND_COMPONENTS : core;base
-- F:/msys2/mingw64/share/cmake/Modules/FindwxWidgets.cmake(176): wxWidgets_SELECT_OPTIONS=--debug=no;--static=no
-- F:/msys2/mingw64/share/cmake/Modules/FindwxWidgets.cmake(180): wxWidgets_CXX_FLAGS=-IF:/msys2/mingw64/lib/wx/include/msw-unicode-3.1 -IF:/msys2/mingw64/include/wx-3.1 -D_FILE_OFFSET_BITS=64 -DWXUSINGDLL -D__WXMSW__ -fpermissive
-- F:/msys2/mingw64/share/cmake/Modules/FindwxWidgets.cmake(180): wxWidgets_DEFINITIONS=_FILE_OFFSET_BITS=64;WXUSINGDLL;__WXMSW__
-- F:/msys2/mingw64/share/cmake/Modules/FindwxWidgets.cmake(180): wxWidgets_INCLUDE_DIRS=F:/msys2/mingw64/lib/wx/include/msw-unicode-3.1;F:/msys2/mingw64/include/wx-3.1
-- F:/msys2/mingw64/share/cmake/Modules/FindwxWidgets.cmake(180): wxWidgets_CXX_FLAGS=-fpermissive
-- F:/msys2/mingw64/share/cmake/Modules/FindwxWidgets.cmake(180): wxWidgets required components : core;base
-- F:/msys2/mingw64/share/cmake/Modules/FindwxWidgets.cmake(180): wxWidgets optional components :
-- F:/msys2/mingw64/share/cmake/Modules/FindwxWidgets.cmake(180): wxWidgets_LIBRARIES=-LF:/msys2/mingw64/lib;;;-pipe;-Wl,--dynamicbase,--high-entropy-va,--nxcompat,--default-image-base-high;-Wl,--subsystem,windows;-mwindows;-lwx_mswu_core-3.1;-lwx_baseu-3.1
-- F:/msys2/mingw64/share/cmake/Modules/FindwxWidgets.cmake(180): wxWidgets_LIBRARY_DIRS=F:/msys2/mingw64/lib
-- F:/msys2/mingw64/share/cmake/Modules/FindwxWidgets.cmake(176): _wx_filename:  F:/msys2/mingw64/include/wx-3.1/wx/version.h
-- F:/msys2/mingw64/share/cmake/Modules/FindwxWidgets.cmake(176): wxWidgets_VERSION_STRING:    3.1.5
-- F:/msys2/mingw64/share/cmake/Modules/FindwxWidgets.cmake(176): wxWidgets_FOUND           : TRUE
-- F:/msys2/mingw64/share/cmake/Modules/FindwxWidgets.cmake(176): wxWidgets_INCLUDE_DIRS    : F:/msys2/mingw64/lib/wx/include/msw-unicode-3.1;F:/msys2/mingw64/include/wx-3.1
-- F:/msys2/mingw64/share/cmake/Modules/FindwxWidgets.cmake(176): wxWidgets_LIBRARY_DIRS    : F:/msys2/mingw64/lib
-- F:/msys2/mingw64/share/cmake/Modules/FindwxWidgets.cmake(176): wxWidgets_LIBRARIES       : -LF:/msys2/mingw64/lib;;;-pipe;-Wl,--dynamicbase,--high-entropy-va,--nxcompat,--default-image-base-high;-Wl,--subsystem,windows;-mwindows;-lwx_mswu_core-3.1;-lwx_baseu-3.1
-- F:/msys2/mingw64/share/cmake/Modules/FindwxWidgets.cmake(176): wxWidgets_CXX_FLAGS       : -fpermissive
-- F:/msys2/mingw64/share/cmake/Modules/FindwxWidgets.cmake(176): wxWidgets_USE_FILE        : F:/msys2/mingw64/share/cmake/Modules/UsewxWidgets.cmake
rc file=/include/wx/msw/wx.rc
-- Configuring done
-- Generating done
-- Build files have been written to: E:/code/wxOpenCVCameras/build
~~~~

Then running `mingw32-make` works OK.

This method is mentioned in this discussion's post here:  https://github.com/msys2/MINGW-packages/issues/10963#issuecomment-1113266308

However, this `export` method works under the mitty shell of msys2, it does not work under normal Windows command line.

Under the later command line, we have to use `set` command to set a environment variable. (see a manual here: [Environment Variables for Java Applications - PATH, CLASSPATH, JAVA_HOME](https://www3.ntu.edu.sg/home/ehchua/programming/howto/Environment_Variables.html) )

**EDIT: 2022-05-04**
I try to set the environment variable under Windows Command Line, by the `set` command, when you try to run the cmake.exe comand to generate the makefiles. You have still one tweak on the file `FindwxWidgets.cmake`, see my updated tutorial here:  https://github.com/msys2/MINGW-packages/issues/11585#issuecomment-1113904778

Either run command
~~~~
E:\code\wxOpenCVCameras\build>set WX_CONFIG=F:\msys2\mingw64/bin/wx-config-3.1
~~~~

Or add a line in your CMakefiles.txt
~~~~
set(wxWidgets_CONFIG_EXECUTABLE "wx-config-3.1")
~~~~

**Comment by PBfordev** [comment]
My point was that I believe that a proper CMakefile should not assume anything about the build system, folder structure, or the libraries.

It should work on any properly set-up system having supported compilers and libraries.

The original CMakefile works out of the box both with MinGW and Microsoft compilers (and likely also with clang) on Windows. IIIRC, it worked also on Linux with GCC for me.

I do understand it can get complicated with a nonstandard environment such as MSYS2, in particular when multiple libraries versions can be supported.

Be that as it may, wxOpenCVCameras is not a real application or library, it is just a proof of concept code demonstration, so I do not think much effort needs to be spent on building it.

**Comment by asmwarrior** [comment]
OK, I understand your comments.

I see another issue about this application(when several net videos are playing in thumbnail mode), when I close the application, it always get hanged.

**Comment by PBfordev** [comment]
> I see another issue about this application(when several net videos are playing in thumbnail mode), when I close the application, it always get hanged.

I have never experienced this. Sometimes, it may take a while (see Notes) but just a couple of seconds at most and then the application quits.

If it really does hang, you need to use a debugger to know where it does.

**Comment by PBfordev** [comment]
I have tried to reproduce it with the latest MSYS2 packages [mingw-w64-x86_64-opencv](https://packages.msys2.org/package/mingw-w64-x86_64-opencv?repo=mingw64) and https://packages.msys2.org/group/mingw-w64-x86_64-toolchain; as well as self-built Release Shared version of wxWidgets master.

I launched the application, pressed <Ctrl+A> and waited until all cameras started receiving. I could not connect to "Bunny 2" but I do not think it matters here.

After all cameras have been capturing for some time, I clicked the close button in the main frame caption. Most of the cameras closed immediately, there was one empty camera panel still displayed for a couple of seconds while the caption indicated the application is not responding. This cannot be helped, see the Notes (unless one is willing to kill a thread, which I do not recommend). In a real application, I would hide the frame before closing it, to not scare the user.
